src/core/expr/fexpr_fillna.cc

```diff
@@ -89,55 +103,107 @@ class FExpr_FillNA : public FExpr_Func {
 
     Workframe evaluate_n(EvalContext &ctx) const override {
       Workframe wf = arg_->evaluate_n(ctx);
-      Groupby gby = ctx.get_groupby();
-      if (!gby) {
-        gby = Groupby::single_group(wf.nrows());
-      } else {
-        wf.increase_grouping_mode(Grouping::GtoALL);
-      }
-
-      for (size_t i = 0; i < wf.ncols(); ++i) {
-        bool is_grouped = ctx.has_group_column(
-                            wf.get_frame_id(i),
-                            wf.get_column_id(i)
-                          );
-        if (is_grouped) continue;
-
-        Column coli = wf.retrieve_column(i);
-        auto stats = coli.get_stats_if_exist();
-        bool na_stats_exists = stats && stats->is_computed(Stat::NaCount);
-        bool has_nas = na_stats_exists? stats->nacount()
-                                      : true;
-
-        if (has_nas) {
-          RowIndex ri = reverse_? fill_rowindex<true>(coli, gby)
-                                : fill_rowindex<false>(coli, gby);
-          coli.apply_rowindex(ri);
+      Workframe wff = arg_->evaluate_n(ctx);
+      bool hasValue = (value_->get_expr_kind() != Kind::None);
+      // scalar fill
+      if (hasValue) {
+        Workframe wf_value = value_->evaluate_n(ctx);
+        if (wf_value.ncols() == 1) wf_value.repeat_column(wf.ncols());
+        if (wf_value.ncols() != wf.ncols()) {
+          throw TypeError() << "Incompatible number of columns in " << repr()
+            << ": the first argument has " << wf.ncols() <<", while the `value` "
+            <<"argument has " << wf_value.ncols();
         }
 
-        wf.replace_column(i, std::move(coli));
-      }
+        wf.sync_grouping_mode(wf_value);
+        Workframe outputs(ctx);
+
+        for (size_t i = 0; i < wf.ncols(); ++i) {
+          Column cond_col = wf.retrieve_column(i);
+          Column orig_col = wff.retrieve_column(i);
+          Column repl_col = wf_value.retrieve_column(i);
+
+          SType out_stype = common_stype(orig_col.stype(), repl_col.stype());
+          repl_col.cast_inplace(out_stype);
+          orig_col.cast_inplace(out_stype);
+          Column cond = evaluate1(std::move(cond_col));
+          Column out = Column{new IfElse_ColumnImpl(std::move(cond), std::move(repl_col), std::move(orig_col))};
+          wff.replace_column(i, std::move(out));
+          }
+
+        return wff;
+
+        } 
+      // forward/backward fill
+       else {
```

Comment:
Something is wrong with the indentation here.

Reply:
do you mind explaining what you mean? two spaces? four?

Reply:
- indentation for the for curly brackets;
- indentation for comment;
- indentaton for `else`.

I have posted an example of how it may look below.